### PR TITLE
Implement Akima1DInterpolator

### LINKED ISSUE #7588: Add Akima Interpolation to Pandas
Currently Akima is not a valid interpolation method in Pandas.  However scipy does have Akima1Dinterpolator.  As Pandas uses SciPy for doing interpolation, adding the Akima method to pandas should be easy.

### PR DESCRIPTION
**Pull request by gliptak**
- [x] closes #7588
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by gliptak** [comment]
This build failed https://travis-ci.org/pydata/pandas/jobs/121843938 likely because of the ```Scipy``` version.

```axis``` was added to ```Akima1DInterpolator``` at ```0.17.0``` as per https://github.com/scipy/scipy/commit/b0395b40c577042436511ce13b9bc067508956db

Is there a minimum version of ```Scipy``` for ```pandas```? What would be a good way to handle lower versions?

Thanks

File: pandas/tests/test_generic.py [hunk 1]
```diff
@@ -747,6 +754,19 @@ def test_interp_regression(self):
...
+        _skip_if_no_akima()
+
+        ser = Series(np.sort(np.random.uniform(size=100)))
+
+        # interpolate at new_index
+        new_index = ser.index.union(Index([49.25, 49.5, 49.75, 50.25, 50.5,
+                                           50.75]))
+        interp_s = ser.reindex(new_index).interpolate(method='akima')
+        # does not blow up, GH5977
+        interp_s[49:51]
```
Comment (jreback) [inline pandas/tests/test_generic.py]: pls have an expected case and use ``assert_series_equal``

**Comment by jreback** [comment]
you can you call this in a try: except to handle the scipy signature differences. (then if it doesn't work in the excpet, raise again). we don't have a min version.

File: pandas/tests/test_generic.py (hunk 1 above)
Comment (gliptak) [inline pandas/tests/test_generic.py]: Added assert

File: pandas/core/missing.py [hunk 2]
```diff
@@ -267,6 +274,56 @@ def _interpolate_scipy_wrapper(x, y, new_x, method, fill_value=None,
...
+    -------
+    y : scalar or array_like
+        The result, of length R or length M or M by R,
+
+    """
+    from scipy import interpolate
+    try:
+        P = interpolate.Akima1DInterpolator(xi, yi, axis=axis)
+    except TypeError:
+        # Scipy earlier than 0.17.0 missing axis
```
Comment (jreback) [inline pandas/core/missing.py]: so need a test for this. on the 2.7_LOCALE build we install a really old scipy. We *may* need to install a somewhat newer one on another build to explicity test for this. When did this ``Akima1DInterpolate`` appear in scipy? and seems the axis parm was added in 0.17.0?

File: pandas/tests/test_generic.py [hunk 3]
```diff
@@ -747,6 +754,20 @@ def test_interp_regression(self):
...
+        _skip_if_no_akima()
+
+        ser = Series([10, 11, 12, 13])
+
+        expected = Series([11.00, 11.25, 11.50, 11.75, 12.00, 12.25, 12.50, 12.75, 13.00],
+                          index=Index([1.0, 1.25, 1.5, 1.75, 2.0, 2.25, 2.5, 2.75, 3.0]))
+        # interpolate at new_index
+        new_index = ser.index.union(Index([1.25, 1.5, 1.75, 2.25, 2.5, 2.75]))
+        interp_s = ser.reindex(new_index).interpolate(method='akima')
+        # does not blow up, GH5977
```
Comment (jreback) [inline pandas/tests/test_generic.py]: give a nicer message then ``does not blow up``

**Comment by gliptak** [comment]
The two test failures seem to be unrelated.

File: pandas/tests/test_generic.py (hunk 3 above)
Comment (gliptak) [inline pandas/tests/test_generic.py]: That is what in ```test_interp_regression_pchip()```. I removed the comment.

File: pandas/core/missing.py (hunk 2 above)
Comment (gliptak) [inline pandas/core/missing.py]: 0.14.0 (as per https://github.com/scipy/scipy/commit/a0a3b382582a1c493ed2e223dfdf9ca835f9c7cd) I implemented a different check.

**Comment by gliptak** [comment]
The two test failures seem to be unrelated.

File: pandas/core/missing.py [hunk 4]
```diff
@@ -236,9 +236,17 @@ def _interpolate_scipy_wrapper(x, y, new_x, method, fill_value=None,
         alt_methods['pchip'] = interpolate.pchip_interpolate
     except AttributeError:
         if method == 'pchip':
-            raise ImportError("Your version of scipy does not support "
+            raise ImportError("Your version of Scipy does not support "
                               "PCHIP interpolation.")
 
+    try:
```
Comment (jreback) [inline pandas/core/missing.py]: can you change this (and the pchip one to something like):

```
if method == 'akima':
    alt_methods['akima'] = _akima_interpolate
```

Move to ``_akima_interpolate``
```
    from scipy.interpolate import Akima1DInterpolator  # noqa
```

File: pandas/core/missing.py [hunk 5]
```diff
@@ -267,6 +275,56 @@ def _interpolate_scipy_wrapper(x, y, new_x, method, fill_value=None,
...
+    --------
+    Akima1DInterpolator
+
+    Returns
+    -------
+    y : scalar or array_like
+        The result, of length R or length M or M by R,
+
+    """
+    from scipy import interpolate
```
Comment (jreback) [inline pandas/core/missing.py]: to here

**Comment by jreback** [comment]
pls add a whatsnew as well

File: pandas/core/missing.py (hunk 4 above)
Comment (gliptak) [inline pandas/core/missing.py]: Updated.

File: pandas/core/missing.py (hunk 5 above)
Comment (gliptak) [inline pandas/core/missing.py]: I'm not sure what you are asking here ...

**Comment by gliptak** [comment]
Added whatsnew

File: pandas/core/missing.py (hunk 5 above)
Comment (jreback) [inline pandas/core/missing.py]: move the import of the Akima1DInterpolate to ``_akima_interpolate``.

And change how its handled in the ``alt_methods`` above

File: pandas/core/missing.py [hunk 6]
```diff
@@ -232,18 +232,18 @@ def _interpolate_scipy_wrapper(x, y, new_x, method, fill_value=None,
...
             raise ImportError("Your version of Scipy does not support "
                               "PCHIP interpolation.")
 
-    try:
-        from scipy.interpolate import Akima1DInterpolator  # noqa
-        alt_methods['akima'] = _akima_interpolate
-    except ImportError:
-        if method == 'akima':
+    if method == 'akima':
+        try:
```
Comment (jreback) [inline pandas/core/missing.py]: yes this is fine

File: pandas/core/missing.py (hunk 2 above)
Comment (jreback) [inline pandas/core/missing.py]: ok the 3.4 build has ``scipy==0.14.0`` installed so pls verify that the test works corretctly (IOW it is NOT skipped, but passes).

remove ``scipy`` from ``ci/requirements-3.5_OSX``. Make sure that you DO see some skips. I don't think we were testing having NO scipy.

**Comment by jreback** [comment]
pls squash

**Comment by jreback** [comment]
add a small note when to use this in : http://pandas-docs.github.io/pandas-docs-travis/missing_data.html#interpolation (you might want to break up that section a bit and make it into a table)

**Comment by jreback** [comment]
add to the doc-string for ``.interpolate`` with a ``versionadded`` tag

**Comment by gliptak** [comment]
Added notes and versionadded tag

File: pandas/core/missing.py [hunk 7]
```diff
@@ -101,6 +101,7 @@ def interpolate_1d(xvalues, yvalues, method='linear', limit=None,
 
     Bounds_error is currently hardcoded to False since non-scipy ones don't
     take it as an argumnet.
+    .. versionadded:: 0.18.1 support for 'akima'.
```
Comment (jreback) [inline pandas/core/missing.py]: no this needs to be updated in ``pandas.core.generic.NDFrame.interpolate``

Comment (gliptak) [inline pandas/core/missing.py]: Moved section.

File: doc/source/missing_data.rst [hunk 8]
```diff
@@ -394,7 +394,8 @@ You'll want to consult the full scipy interpolation documentation_ and reference
 The appropriate interpolation method will depend on the type of data you are working with.
 For example, if you are dealing with a time series that is growing at an increasing rate,
 ``method='quadratic'`` may be appropriate.  If you have values approximating a cumulative
```
Comment (jreback) [inline doc/source/missing_data.rst]: can you change this to bullet points

File: pandas/core/generic.py [hunk 9]
```diff
@@ -3465,12 +3466,13 @@ def interpolate(self, method='linear', axis=0, limit=None, inplace=False,
...
               e.g. df.interpolate(method='polynomial', order=4).
               These use the actual numerical values of the index.
-            * 'krogh', 'piecewise_polynomial', 'spline', and 'pchip' are all
+            * 'krogh', 'piecewise_polynomial', 'spline', 'pchip' and 'akima' are all
               wrappers around the scipy interpolation methods of similar
               names. These use the actual numerical values of the index. See
               the scipy documentation for more on their behavior
               `here <http://docs.scipy.org/doc/scipy/reference/interpolate.html#univariate-interpolation>`__  # noqa
               `and here <http://docs.scipy.org/doc/scipy/reference/tutorial/interpolate.html>`__  # noqa
+            .. versionadded:: 0.18.1 support for 'akima'.
```
Comment (jreback) [inline pandas/core/generic.py]: need a blank line here above (or it gives warnings)

File: pandas/core/missing.py [hunk 10]
```diff
@@ -232,13 +232,21 @@ def _interpolate_scipy_wrapper(x, y, new_x, method, fill_value=None,
...
-        if method == 'pchip':
-            raise ImportError("Your version of scipy does not support "
+    if method == 'pchip':
+        try:
+            alt_methods['pchip'] = interpolate.pchip_interpolate
+        except AttributeError:
+            raise ImportError("Your version of Scipy does not support "
                               "PCHIP interpolation.")
 
+    if method == 'akima':
```
Comment (jreback) [inline pandas/core/missing.py]: elif

File: pandas/core/missing.py [hunk 11]
```diff
@@ -267,6 +275,56 @@ def _interpolate_scipy_wrapper(x, y, new_x, method, fill_value=None,
...
+        How many derivatives to extract; None for all potentially
+        nonzero derivatives (that is a number equal to the number
+        of points), or a list of derivatives to extract. This number
+        includes the function value as 0th derivative.
+    axis : int, optional
+        Axis in the yi array corresponding to the x-coordinate values.
+
+    See Also
+    --------
+    Akima1DInterpolator
```
Comment (jreback) [inline pandas/core/missing.py]: I think you have to do ``scipy.interpolate.Akima1DInterpolator`` @jorisvandenbossche ? do we even have ``scipy`` in the xrefs?

File: pandas/tests/test_generic.py [hunk 12]
```diff
@@ -734,7 +741,7 @@ def test_interpolate(self):
         non_ts[0] = np.NaN
         self.assertRaises(ValueError, non_ts.interpolate, method='time')
 
-    def test_interp_regression(self):
+    def test_interp_regression_pchip(self):
```
Comment (jreback) [inline pandas/tests/test_generic.py]: just make this ``test_interp_pchip`` (and same below, no need for ``regression``)

**Comment by gliptak** [comment]
Updated per comments.

File: doc/source/missing_data.rst [hunk 13]
```diff
@@ -392,9 +392,12 @@ The ``method`` argument gives access to fancier interpolation methods.
 If you have scipy_ installed, you can set pass the name of a 1-d interpolation routine to ``method``.
 You'll want to consult the full scipy interpolation documentation_ and reference guide_ for details.
 The appropriate interpolation method will depend on the type of data you are working with.
-For example, if you are dealing with a time series that is growing at an increasing rate,
-``method='quadratic'`` may be appropriate.  If you have values approximating a cumulative
-distribution function, then ``method='pchip'`` should work well.
+
+* If you are dealing with a time series that is growing at an increasing rate,
+``method='quadratic'`` may be appropriate.
+* If you have values approximating a cumulative distribution function,
```
Comment (jreback) [inline doc/source/missing_data.rst]: good

File: pandas/core/missing.py (hunk 2 above)
Comment (jreback) [inline pandas/core/missing.py]: ok , just remove ``scipy`` from ``ci/requirements-3.5_OSX`` and I think will be good to go.

Comment (gliptak) [inline pandas/core/missing.py]: from ci/requirements-3.5_OSX.run ?

Comment (jreback) [inline pandas/core/missing.py]: yep (we are including a version of scipy on EVERY build, so need to take it from one)

Comment (gliptak) [inline pandas/core/missing.py]: Done?

File: pandas/core/missing.py (hunk 5 above)
Comment (gliptak) [inline pandas/core/missing.py]: Would you still like to see a change around this?

Comment (jreback) [inline pandas/core/missing.py]: this is ok

**Comment by jreback** [comment]
ok, looks good. ping when green (and pls verify that tests are skipped as appropriate)

File: pandas/core/missing.py (hunk 11 above)
Comment (jorisvandenbossche) [inline pandas/core/missing.py]: Yes, that should indeed work. But you should add scipy to the link mapping here: https://github.com/pydata/pandas/blob/master/doc/source/conf.py#L295 (you can just look how it is done for numpy)

File: doc/source/missing_data.rst [hunk 14]
```diff
@@ -392,9 +392,12 @@ The ``method`` argument gives access to fancier interpolation methods.
 If you have scipy_ installed, you can set pass the name of a 1-d interpolation routine to ``method``.
 You'll want to consult the full scipy interpolation documentation_ and reference guide_ for details.
 The appropriate interpolation method will depend on the type of data you are working with.
-For example, if you are dealing with a time series that is growing at an increasing rate,
-``method='quadratic'`` may be appropriate.  If you have values approximating a cumulative
-distribution function, then ``method='pchip'`` should work well.
+
+* If you are dealing with a time series that is growing at an increasing rate,
+``method='quadratic'`` may be appropriate.
```
Comment (jorisvandenbossche) [inline doc/source/missing_data.rst]: You need two spaces here at the beginning of this line, so the start of the text is aligned with the start of the text in the previous line ('If ...')

File: doc/source/missing_data.rst [hunk 15]
```diff
@@ -392,9 +392,12 @@ The ``method`` argument gives access to fancier interpolation methods.
...
+* If you have values approximating a cumulative distribution function,
+then ``method='pchip'`` should work well.
```
Comment (jorisvandenbossche) [inline doc/source/missing_data.rst]: the same here

File: pandas/core/generic.py [hunk 16]
```diff
@@ -3465,13 +3466,15 @@ def interpolate(self, method='linear', axis=0, limit=None, inplace=False,
...
               These use the actual numerical values of the index.
-            * 'krogh', 'piecewise_polynomial', 'spline', and 'pchip' are all
+            * 'krogh', 'piecewise_polynomial', 'spline', 'pchip' and 'akima' are all
               wrappers around the scipy interpolation methods of similar
               names. These use the actual numerical values of the index. See
               the scipy documentation for more on their behavior
               `here <http://docs.scipy.org/doc/scipy/reference/interpolate.html#univariate-interpolation>`__  # noqa
               `and here <http://docs.scipy.org/doc/scipy/reference/tutorial/interpolate.html>`__  # noqa
 
+            .. versionadded:: 0.18.1 support for 'akima'.
```
Comment (jorisvandenbossche) [inline pandas/core/generic.py]: The 'support for 'akima'' should go on the next line I think. See the sphinx docs: http://www.sphinx-doc.org/en/stable/markup/para.html#directive-versionadded for an example. And then maybe change the text to something like "Added support for the 'akima' method"

**Comment by jorisvandenbossche** [comment]
@gliptak I just added some more doc comments, looks very good for the rest!

File: pandas/core/generic.py (hunk 16 above)
Comment (gliptak) [inline pandas/core/generic.py]: Changed.

File: doc/source/missing_data.rst (hunk 14 above)
Comment (gliptak) [inline doc/source/missing_data.rst]: Done

File: doc/source/missing_data.rst (hunk 15 above)
Comment (gliptak) [inline doc/source/missing_data.rst]: Done

File: pandas/core/missing.py (hunk 11 above)
Comment (gliptak) [inline pandas/core/missing.py]: Done

**Comment by gliptak** [comment]
@jreback Build is green. Tests skipped https://s3.amazonaws.com/archive.travis-ci.org/jobs/122101334/log.txt

```
#268 pandas.tests.test_generic.TestSeries.test_interpolate_akima: no scipy.stats module
#269 pandas.tests.test_generic.TestSeries.test_interpolate_corners: no scipy.stats module
#270 pandas.tests.test_generic.TestSeries.test_interpolate_pchip: no scipy.stats module
```

I also opened https://github.com/pydata/pandas/issues/12856 (consistently fails the same way)

**Comment by jreback** [comment]
@gliptak thanks!

pls review the built docs for accuracy (and good looking ness) and issue a follow up if necessary. (prob will take a few hours before they are built)

**Comment by gliptak** [comment]
@jreback @jorisvandenbossche Thank you for guiding this through

**Comment by gliptak** [comment]
Docs updated good.

**Pull request closed without merging**